xgboost_ray/main.py

```diff
@@ -419,39 +524,67 @@ def _train(params: Dict,
     else:
         params["nthread"] = cpus_per_actor
 
-    # Create queue for communication from worker to caller.
-    # Always create queue.
-    queue = Queue()
-
     # Create remote actors
-    actors = [
-        _create_actor(
+    def handle_actor_failure(actor_id):
+        rank = _actors.index(actor_id)
+        _failed_actor_ranks.add(rank)
+        _actors[rank] = None
+
+    newly_created = 0
+    for i in list(_failed_actor_ranks):
+        if _actors[i] is not None:
+            raise RuntimeError(
+                f"Trying to create actor with rank {i}, but it already "
+                f"exists.")
+        actor = _create_actor(
             rank=i,
             num_actors=ray_params.num_actors,
             num_cpus_per_actor=cpus_per_actor,
             num_gpus_per_actor=gpus_per_actor,
             resources_per_actor=ray_params.resources_per_actor,
-            queue=queue,
+            queue=_queue,
             checkpoint_frequency=ray_params.checkpoint_frequency)
-        for i in range(ray_params.num_actors)
+        _actors[i] = actor
+        # Remove from this set so it is not created again
+        _failed_actor_ranks.remove(i)
+        newly_created += 1
+
+    # Maybe we got a new queue actor
+    wait_queue = [
+        actor.set_queue.remote(_queue) for actor in _actors
+        if actor is not None
```

Comment:
can there actually be `actor is None` here?

Reply:
Yes! After an actor failed its entry in `_actors` is set to None in the `train()` function. `_train` is then called again, so entries can be None

Reply:
Hmm, don't all the None entries get filled in L586 to 603?

Reply:
Only those that are in `_failed_actor_ranks` - and this set is empty if we continue with fewer actors. See previous comment below